src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+{-# OPTIONS --safe #-}
+-- {-# OPTIONS -v tactic.premises:100 #-}
+module Tactic.Premises where
+
+open import Prelude
+import Data.List as L
+open import Data.Fin using (toℕ)
+
+open import Class.Show.Core using (show; Show)
+open import Class.Show.Instances using ()
+open import Class.Semigroup.Core; open import Class.Semigroup.Instances
+open import Class.Monad.Core; open import Class.Monad.Instances
+open import Class.Applicative.Core; open import Class.Applicative.Instances
+
+open import Interface.Functor using (map; _<$>_)
+open import Interface.Bifunctor using (map₁; map₂)
+open import Interface.DecEq.Ext
+open import Interface.Decidable.Instance
+open import Interface.STS
+
+open import Generics; open Debug ("tactic.premises", 100)
+open import Generics.Utils.Ext
+open import Agda.Builtin.Reflection using (onlyReduceDefs; getInstances)
+open import Reflection hiding (_>>=_; _>>_; _≟_)
+open import Reflection.AST.Term hiding (_≟_)
+open import Reflection.AST.Abstraction using (unAbs)
+
+{-
+** Extracting the hypotheses of an STS rule.
+
+Constructors should be formulated in the STS syntax for inference rules, i.e.
+```
+  rule : ∀ <IMPLICIT_ARGUMENTS + LET_EXPRESSIONS>
+    ∙ <HYPOTHESIS#1>
+    ∙ <HYPOTHESIS#2>
+    ⋮
+    ∙ <HYPOTHESIS#n-1>
+    ∙ <HYPOTHESIS#n>
+      ────────────────────────────────
+      <CONCLUSION>
+```
+This meta-program then generates the type of the (decidable) hypotheses of the rule,
+as well as a proof that it is in fact decidable, i.e.
+```
+rule-premises : ∀ ⋯ → Set
+rule-premises ⋯ =
+  ∙ <HYPOTHESIS#1>
+  ∙ <HYPOTHESIS#3>
+  ⋮
+  ∙ <HYPOTHESIS#n-1>
+
+rule-premises? : ∀ ⋯ → Dec (rule-premises ⋯)
+rule-premises? = it
+```
+where ⋯ stands for the subset of implicit arguments that are relevant for the
+selected subset of decidable hypotheses.
+
+To be precise, we generate both definitions above under a single name, i.e.
+```
+unquoteDecl rule-premises = genPremises rule-premises (quote rule)
+-- rule-premises        : ∀ ⋯ → ∃ Dec
+-- rule-premises .proj₁ : ∀ ⋯ → Set
+-- rule-premises .proj₂ : ∀ ⋯ → Dec (rule-premises .proj₁ ⋯)
+```
+-}
+genPremises : Name → Name → TC ⊤
+genPremises f n = do
+  print $ "*** Generating premises for constructor:" <+> show n <+> "***"
+  ty ← reduceRuleSyntax =<< getType n                  -- (1)
+  let is , hs = breakImplicits $ viewTy ty .proj₁      -- (2)
+  hs′ ← map unbundleHypotheses                         -- (3)
+      $ extendContextTel (absTelescope is)
+      $ mapM reduceRuleSyntax hs
+  dhs ← forM hs′ $ λ h → (_, h) <$> isDecidable? is h  -- (4)
+  let dhs′ = proj₂ <$> filter ((_≟ true) ∘ proj₁) dhs
```

Comment:
This line and the one above are really just a monadic `filter` function:
```agda

module _ {A : Set} {M : Type↑} ⦃ _ : Functor M ⦄ ⦃ _ : Monad M ⦄ where
  filterM : (A → M Bool) → List A → M (List A)
  filterM P? [] = return []
  filterM P? (x ∷ xs) = do
    b ← P? x
    xs ← filterM P? xs
    return $ if b then x ∷ xs else xs
```

If you use that, then you can just write `let dhs′ = filterM (isDecidable? is) hs`, which I think is a lot clearer.